Feature: Add impersonation to tekton results API

sayan-biswas opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Resolves #309 

This feature adds Kubernetes impersonation header processing, impersonation access check for requester and RBAC for tekton results resources with impersonated user data. 

https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation

<!-- /kind bug, cleanup, design, documentation, feature, flake, misc, question, tep -->
# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you review them:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [Tested your changes locally](https://github.com/tektoncd/results/blob/main/docs/DEVELOPMENT.md) (if this is a code change)
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user-facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contain the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Kubernetes impersonation headers can be used when authorizing access to Tekton Results APis.

ACTION REQUIRED: The NO_AUTH config parameter was renamed to AUTH_DISABLE
```

sayan-biswas commented:
/kind feature

sayan-biswas commented:
/test all

sayan-biswas commented:
/test pull-tekton-results-integration-tests

sayan-biswas commented:
/assign @adambkaplan

adambkaplan commented on pkg/api/server/v1alpha2/auth/impersonation/impersonation.go:
```diff
@@ -0,0 +1,217 @@
+			return err
+		}
+		if !sar.Status.Allowed {
+			return fmt.Errorf("forbidden: '%s' doesn't have access to impersonate '%s'", requester, resourceAttribute.Resource)
```
small change to error message:

```suggestion
			return fmt.Errorf("forbidden: '%s' doesn't have permission to impersonate '%s'", requester, resourceAttribute.Resource)
```

sayan-biswas replied:
:thumbsup: Changed

adambkaplan commented on pkg/api/server/v1alpha2/auth/rbac.go:
```diff
@@ -89,11 +109,32 @@ func (r *RBAC) Check(ctx context.Context, namespace, resource, verb string) erro
+		groups := []string{"tekton.dev"}
+		extra := map[string]authzv1.ExtraValue{}
+
+		// Check whether the authenticated user has access to impersonate
```
Replace "access" with permission. Admins grant _access_, users receive _permission_.

```suggestion
		// Check whether the authenticated user has permission to impersonate
```

sayan-biswas replied:
:thumbsup: Changed.

adambkaplan commented on pkg/api/server/v1alpha2/auth/rbac.go:
```diff
@@ -38,22 +39,41 @@ import (
 type RBAC struct {
-	authn authnclient.AuthenticationV1Interface
-	authz authzclient.AuthorizationV1Interface
+	impersonation bool
```
Maybe call this `allowImpersonation` to avoid confusion.

sayan-biswas replied:
renamed to `allowImpersonation `

adambkaplan commented on pkg/api/server/v1alpha2/auth/rbac.go:
```diff
@@ -114,3 +155,19 @@ func (r *RBAC) Check(ctx context.Context, namespace, resource, verb string) erro
+}
+
+// WithImpersonation is an option function to enable Impersonation
+func WithImpersonation() Option {
```
With this kind of builder/factory pattern, you should accept an argument so it can be explicitly set to true or false.

```suggestion
func WithImpersonation(enabled bool) Option {
```

sayan-biswas replied:
Added argument

adambkaplan commented on pkg/api/server/v1alpha2/auth/rbac.go:
```diff
@@ -38,22 +39,41 @@ import (
 	}
+
+	// Parse Impersonation header if the feature is enabled
+	var i *impersonation.Impersonation
```
Please use a more descriptive variable name here.

```suggestion
	var impersonator *impersonation.Impersonation
```

sayan-biswas replied:
updated

adambkaplan commented on pkg/api/server/v1alpha2/auth/impersonation/impersonation.go:
```diff
@@ -0,0 +1,217 @@
+		// https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation
+		if hasGroups || hasExtra || hasUID {
+			return ErrorImpersonateUserRequired
+		} else {
+			return ErrorNoImpersonationData
+		}
```
Should we return an error here? This seems to imply that _all_ requests require impersonation, which is certainly not the case for requests that come in from the watcher.

sayan-biswas replied:
I returned error here so that the `NewImpersonation` function doesn't return an empty or nil `Impersonation` struct without returning any error.
This error will not stop the processing of the requests, because it's handled here.

https://github.com/tektoncd/results/blob/06adf689eb14324be872c43787d6d68861df5af9/pkg/api/server/v1alpha2/auth/rbac.go#L71-L74

The idea here is if no impersonation header is present we, continue processing with a nil `Impersonation` struct, which is checked later while performing impersonation check. But if any other impersonation header is present without having `Impersonate-User` header, then we stop processing of that request. This is the same behaviour in the Kubernetes impersonation filter implementation.

adambkaplan commented on config/base/env/config:
```diff
@@ -10,7 +10,8 @@ REST_PORT=8080
 TLS_HOSTNAME_OVERRIDE=
 TLS_PATH=/etc/tls
-NO_AUTH=false
+AUTH_DISABLE=false
```
This absolutely needs to be called out as a breaking change (if we decide to accept it).

sayan-biswas replied:
I think this name makes it easier to categorise the options in the `config`, prepend with a short type description for the option.

adambkaplan commented on config/base/env/config:
```diff
@@ -10,7 +10,8 @@ REST_PORT=8080
 TLS_PATH=/etc/tls
-NO_AUTH=false
+AUTH_DISABLE=false
+AUTH_IMPERSONATE=false
```
Personally, I think we should set the default config to `true` in our release. At minimum this will ensure that we run end to end tests with this enabled, reducing our test matrix.

sayan-biswas replied:
Default set to `true`.

adambkaplan commented on cmd/api/main.go:
```diff
@@ -186,7 +193,7 @@ func main() {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	mux := runtime.NewServeMux()
+	mux := runtime.NewServeMux(runtime.WithIncomingHeaderMatcher(impersonation.HeaderMatcher))
```
Should this be added if the impersonation check is disabled?

sayan-biswas replied:
The default header matcher is implicit in `runtime.NewServeMux()`. This custom function is added so that the requester can send impersonation header with name `Impersonate-User` instead of GRPC gateway header format, `Grpc-Metadata-Impersonate-User`.

But I'll still put in this in a condition.

sayan-biswas replied:
https://github.com/tektoncd/results/blob/06adf689eb14324be872c43787d6d68861df5af9/cmd/api/main.go#L109-L112

adambkaplan reviewed (commented):
/approve

Marking this as an approved feature to add. In the current form, however, we need to ensure that e2e tests work with this feature enabled. @dibyom do other projects have a mechanism to test features e2e where the feature flag is enabled _and_ disabled?

adambkaplan reviewed (commented):
Almost there! Code looks good, I added a suggestion to add a code comment.

Can you also please add a section to the Authorization doc that indicates we support Kubernetes impersonation headers [1]? You don't need to explain what user impersonation is - a link to the Kubernetes doc [2] is sufficient. Just focus on how to enable the feature and how Results clients (gRPC, REST) can use it.

[1] https://github.com/tektoncd/results/tree/main/docs/api#authenticationauthorization
[2] https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation

sayan-biswas commented on pkg/api/server/v1alpha2/auth/rbac.go:
```diff
@@ -38,22 +39,41 @@ import (
+	var err error
+	if r.allowImpersonation {
+		impersonator, err = impersonation.NewImpersonation(md)
+		if err != nil && err != impersonation.ErrorNoImpersonationData {
+			log.Println(err)
+			return status.Error(codes.Unauthenticated, "invalid impersonation data")
+		}
```
Done

adambkaplan commented:
/assign @dibyom @alan-ghelardi 

There is a breaking configuration change here.

khrm commented:
Shouldn't we introduce a deprecation notice instead of directly breaking it? We can remove the `NO_AUTH` in the next release with the ACTION required notice. In this release, we can have a deprecation notice.

dibyom commented on config/base/env/config:
```diff
@@ -10,7 +10,8 @@ REST_PORT=8080
 TLS_PATH=/etc/tls
-NO_AUTH=false
+AUTH_DISABLE=false
+AUTH_IMPERSONATE=true
```
Is this enabled by default? Doc above states that it is disabled by default?

sayan-biswas replied:
Ohh.. I missed this. 
Updated doc.

dibyom commented:
> Shouldn't we introduce a deprecation notice instead of directly breaking it? We can remove the NO_AUTH in the next release with the ACTION required notice. In this release, we can have a deprecation notice.

Technically, we are allowed to make breaking changes in alpha apis. Though generally what we have done in other projects is what @khrm is suggesting here

sayan-biswas commented:
> Shouldn't we introduce a deprecation notice instead of directly breaking it? We can remove the `NO_AUTH` in the next release with the ACTION required notice. In this release, we can have a deprecation notice.

Agreed! But this flag is mostly for development, to bypass the RBAC processing.
Even if this flag is missing from the config, the parses will assume the value as false and continue. Which is the default behaviour anyways. 
Also the release will ship the config file which will anyways have the modified parameter.

sayan-biswas commented:
> Almost there! Code looks good, I added a suggestion to add a code comment.
> 
> Can you also please add a section to the Authorization doc that indicates we support Kubernetes impersonation headers [1]? You don't need to explain what user impersonation is - a link to the Kubernetes doc [2] is sufficient. Just focus on how to enable the feature and how Results clients (gRPC, REST) can use it.
> 
> [1] https://github.com/tektoncd/results/tree/main/docs/api#authenticationauthorization [2] https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation

Added code comment and updated doc.

adambkaplan commented:
@alan-ghelardi just want to check with you that renaming the `NO_AUTH` config setting to `AUTH_DISABLE` is acceptable (since this feature was your contribution).

Otherwise LGTM.
